File: Go/prim_algorithm.go
package main

import (
	"fmt"
	"math"
)

type Edge struct {
	node   int
	weight int
}

func prim(graph map[int][]Edge, n int) int {
	visited := make([]bool, n)
	cost := 0
	visited[0] = true

	for i := 1; i < n; i++ {
		minWeight := math.MaxInt32
		node := -1
		for u := 0; u < n; u++ {
			if visited[u] {
				for _, edge := range graph[u] {
					if !visited[edge.node] && edge.weight < minWeight {
						minWeight = edge.weight
						node = edge.node
					}
				}
			}
		}
		if node != -1 {
			visited[node] = true
			cost += minWeight
		}
	}
	return cost
}

func main() {
	graph := map[int][]Edge{
		0: {{1, 4}, {2, 3}},
		1: {{0, 4}, {2, 1}, {3, 2}},
		2: {{0, 3}, {1, 1}, {3, 4}},
		3: {{1, 2}, {2, 4}},
	}
	fmt.Println("Minimum Cost: ", prim(graph, 4))
}

File: Swift/PrimAlgorithm.swift
import Foundation

struct Edge {
    let node: Int
    let weight: Int
}

func prim(graph: [Int: [Edge]], n: Int) -> Int {
    var visited = Array(repeating: false, count: n)
    var cost = 0
    visited[0] = true

    for _ in 1..<n {
        var minWeight = Int.max
        var nextNode = -1
        for u in 0..<n where visited[u] {
            for edge in graph[u]! where !visited[edge.node] && edge.weight < minWeight {
                minWeight = edge.weight
                nextNode = edge.node
            }
        }
        if nextNode != -1 {
            visited[nextNode] = true
            cost += minWeight
        }
    }
    return cost
}

let graph: [Int: [Edge]] = [
    0: [Edge(node: 1, weight: 4), Edge(node: 2, weight: 3)],
    1: [Edge(node: 0, weight: 4), Edge(node: 2, weight: 1), Edge(node: 3, weight: 2)],
    2: [Edge(node: 0, weight: 3), Edge(node: 1, weight: 1), Edge(node: 3, weight: 4)],
    3: [Edge(node: 1, weight: 2), Edge(node: 2, weight: 4)]
]

print("Minimum Cost: \(prim(graph: graph, n: 4))")
